core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,541 @@
+
+            # Handling exploration reverts.
+            if revert_to_version:
+                old_exploration_stats = exploration_stats_list[
+                    revert_to_version - 1]
+                num_starts = old_exploration_stats['num_starts_v1']
+                num_completions = old_exploration_stats['num_completions_v1']
+                num_actual_starts = old_exploration_stats[
+                    'num_actual_starts_v1']
+                state_stats_mapping = {
+                    state_name: stats_domain.StateStats.from_dict(
+                        old_exploration_stats['state_stats_mapping'][
+                            state_name])
+                    for state_name in old_exploration_stats[
+                        'state_stats_mapping']
+                }
+
+            # Computing num_starts and num_completions.
+            if str(version) in start_exploration_counts_by_version:
+                num_starts += start_exploration_counts_by_version[str(version)]
+            if str(version) in complete_exploration_counts_by_version:
+                num_completions += complete_exploration_counts_by_version[
+                    str(version)]
+
+            # Compute total_hit_count and first_hit_count for the states.
+            for state_name in state_hit_counts_for_this_version:
+                state_stats_mapping[state_name].total_hit_count_v1 += (
+                    state_hit_counts_for_this_version[state_name][
+                        'total_hit_count'])
+                state_stats_mapping[state_name].first_hit_count_v1 += (
+                    state_hit_counts_for_this_version[state_name][
+                        'first_hit_count'])
+
+            # Compute num_completions for the states.
+            for session_id in (
+                    session_id_state_hit_event_mapping_for_this_version):
+                state_hit_event_dicts_sessioned = (
+                    session_id_state_hit_event_mapping_for_this_version[
+                        session_id])
+                # Sort list of state hit events in increasing order of
+                # timestamp.
+                state_hit_event_dicts_sessioned = sorted(
+                    state_hit_event_dicts_sessioned, key=lambda e: e[
+                        'created_on'])
+                # All state hit events in the above list are ordered by their
+                # timestamp. This implies that the last state hit event is the
+                # state at which the learner quit the exploration (This is only
+                # True if the last state hit event is not the terminal state).
+                # So, if the last state is not the terminal state, we increment
+                # num_completions for all states except the last state.
+                # Otherwise, we increment num_completions for all the states.
+                for state_hit_event in state_hit_event_dicts_sessioned[:-1]:
```

Comment:
I don't think I understand why this can't be optimized away. If you create a dict keyed by version whose values are dicts keyed by state name, couldn't you just increment the inner values when walking along the state hit event dicts the first time?

Also, separately, I guess you want to keep track of the event with the largest timestamp. Why not do that in a dict keyed by version and session ID? The value of the dict can be the state name containing the largest timestamp found so far. When iterating along the full list of events the first time, you can just update that value if needed.

Reply:
@seanlip I've got a small issue with this idea. So, our plan is to increment num_completions for all state hit events within a session except the last one (that is where the learner quit).
Now, in what order are events present in values_stringified? When we are doing the first pass over all values, we don't really have any ordering to the events in it right? So, any state hit event we encounter, we won't really know if it is the state hit event with the largest timestamp, would we (since there could be a later event with a larger timestamp)? That was why I thought we had to process all events once, and then sort them again to compute num_completions correctly.

An alternative could be to just increment num_completions for all the state hit events we encounter the first time while maintaining a data structure to keep track of the largest timestamp keyed by session. Then, we can just run a pass through all sessions to find that state hit event with largest timestamp and decrement num_completions for all those values. Would this work?

Reply:
It's true that initially we don't have any ordering. But all you want to do with the ordering is to keep track of which event has the largest timestamp, and the method I proposed accounts for that. I.e. you keep track of the state name corresponding to the largest timestamp on a rolling basis as you make your first pass through the list, updating it as you get more info.

It's a bit like finding the max value in a list by iterating through the list and storing the current max value seen so far, and updating that stored variable only when you find a larger value.

Reply:
@seanlip Yep. I agree that we can find the largest timestamp event on a rolling basis on the first pass. But, from that point, computing num_completions will still be the same complexity. Since our data structure is a dict keyed by version its value is a dict keyed by sessionID, we'll have to iterate over all sessionIDs for a version to increment num_completions anyway. Do I seem to be missing something?

Reply:
@seanlip So, I've figured another way out of this. Instead of doing any of these computations inside the for loop, I did it at the outset. This still involves another pass over the data, but this is just one extra pass and nothing inside the nested for loop. How does it look now?